source/application/wintaser.cpp

```diff
@@ -2885,177 +2874,79 @@ LPCWSTR ExceptionCodeToDescription(DWORD code)
+    }
 }
 
-void RegisterModuleInfo(LPVOID hModule, HANDLE hProcess, LPCWSTR path)
+static void ModuleUnloaded(DWORD process_id, const std::wstring& path)
 {
-	int trusted = IsPathTrusted(path);
-	if(!trusted)
-		return; // we only want to keep track of the trusted ones
+    auto& proc_modules = trusted_modules[process_id];
+    auto removed = std::remove_if(proc_modules.begin(), proc_modules.end(),
+                                  [&path](const std::wstring& m) { return (m.compare(path) == 0); });
 
-	MyMODULEINFO mmi;
-//	if(!GetModuleInformation(hProcess, (HMODULE)hModule, &mmi.mi, sizeof(mmi.mi)))
-	{
-		// GetModuleInformation never works, and neither does EnumProcessModules.
-		// I think it's because we're calling it too early.
-		// (either that or some versions of PSAPI.DLL are simply busted)
-		// so this is the best fallback I could come up with
-		mmi.mi.lpBaseOfDll = hModule;
-		mmi.mi.SizeOfImage = CalculateModuleSize(hModule, hProcess);
-		mmi.mi.EntryPoint = hModule; // don't care
-	}
-	mmi.path = path;
-	if(trusted == 2)
-	{
-		injectedDllModuleInfo = mmi;
-		//SendTrustedRangeInfos(hProcess); // disabled because it's probably not ready to receive it yet.
-	}
-	else
-	{
-		for(unsigned int i = 0; i < trustedModuleInfos.size(); i++)
-		{
-			MyMODULEINFO& mmi2 = trustedModuleInfos[i];
-			if((DWORD)mmi.mi.lpBaseOfDll >= (DWORD)mmi2.mi.lpBaseOfDll
-			&& (DWORD)mmi.mi.lpBaseOfDll+mmi.mi.SizeOfImage <= (DWORD)mmi2.mi.lpBaseOfDll+mmi2.mi.SizeOfImage)
-			{
-				debugprintf(L"apparently already TRUSTED MODULE 0x%08X - 0x%08X (%s)\n", mmi.mi.lpBaseOfDll, (DWORD)mmi.mi.lpBaseOfDll+mmi.mi.SizeOfImage, path);
-				return;
-			}
-		}
-		trustedModuleInfos.push_back(mmi);
-		SendTrustedRangeInfos(hProcess);
-	}
-	debugprintf(L"TRUSTED MODULE 0x%08X - 0x%08X (%s)\n", mmi.mi.lpBaseOfDll, (DWORD)mmi.mi.lpBaseOfDll+mmi.mi.SizeOfImage, path);
+    if (removed == proc_modules.end())
+    {
+        return;
+    }
+    proc_modules.erase(removed, proc_modules.end());
 }
-void UnregisterModuleInfo(LPVOID hModule, HANDLE hProcess, LPCWSTR path)
-{
-	int trusted = IsPathTrusted(path);
-	if(!trusted)
-		return; // we only want to keep track of the trusted ones
-
-	MEMORY_BASIC_INFORMATION mbi = {};
-	if(VirtualQueryEx(hProcess, (char*)hModule+0x1000, &mbi, sizeof(mbi)))
-		hModule = (HMODULE)mbi.AllocationBase;
 
-	if(trusted == 2)
-	{
-		injectedDllModuleInfo.path.clear();
-		memset(&injectedDllModuleInfo.mi, 0, sizeof(injectedDllModuleInfo.mi));
-		//SendTrustedRangeInfos(hProcess); // disabled because the thing we'd send it to is what just got unloaded
-	}
-	else
-	{
-		int count = trustedModuleInfos.size();
-		for(int i = 0; i < count; i++)
-		{
-			if(trustedModuleInfos[i].mi.lpBaseOfDll == hModule)
-			{
-				trustedModuleInfos.erase(trustedModuleInfos.begin() + i);
-				SendTrustedRangeInfos(hProcess);
-				break;
-			}
-		}
-	}
+static bool IsTrustedModule(DWORD process_id, const std::wstring& path)

```

Comment:
Maybe all this stuff needs to move from `wintaser.cpp` into some `trusted_module.cpp`? I'm not a huge fan of the huge `wintase{e,r}.cpp` files with a lot of functions that could well be grouped into separate files.